api/session.ts

```diff
@@ -0,0 +1,429 @@
+		} else {
+			if (userid.length < 1 || !/[a-z]/.test(userid)) {
+				return ';;Your username must contain at least one letter.';
+			}
+			const userstate = await users.get('*', userid);
+			if (userstate) {
+				if (userstate.banstate >= 100 || ((userstate as any).password && userstate.nonce)) {
+					return ';;Your username is no longer available.';
+				}
+				if (userstate.email?.endsWith('@')) {
+					return ';;@gmail';
+				}
+				return ';';
+			} else {
+				// Unregistered username.
+				userType = '1';
+				if (forceUsertype) userType = forceUsertype;
+				data = user + ',' + userType + ',' + time() + ',' + serverHost;
+				this.updateCookie();
+			}
+		}
+		let splitChallenge: string[] = [];
+		for (const delim of [';', '%7C', '|']) {
+			splitChallenge = challenge.split(delim);
+			if (splitChallenge.length > 1) break;
+		}
+		let challengetoken;
+		if (splitChallenge.length > 1) {
+			challengekeyid = parseInt(splitChallenge[0], 10);
+			challenge = splitChallenge[1];
+			if (splitChallenge[2] && !challengetoken) challengetoken = splitChallenge[2];
+		}
+
+		if (!toID(challenge)) {
+			// Bogus challenge.
+			return ';;Corrupt challenge';
+		}
+		if (challengekeyid < 1) {
+			return ';;This server is requesting an invalid login key. This probably means that either you are not connected to a server, or the server is set up incorrectly.';
+		} else if (Config.compromisedkeys?.includes(challengekeyid)) {
+			// Compromised keys - no longer supported.
+			return `;;This server is using login key ${challengekeyid}, which is no longer supported. Please tell the server operator to update their config.js file.`;
+		} else if (!Config.privatekeys?.[challengekeyid]) {
+			// Bogus key id.
+			return ';;Unknown key ID';
+		} else {
+			// Include the challenge in the assertion.
+			data = (challengeprefix || "") + challenge + ',' + data;
+		}
+		this.updateCookie();
+
+		if (Config.validateassertion) {
+			data = Config.validateassertion.call(this, data, user, ip, server);
+		}
+
+		const sign = crypto.createSign('RSA-SHA1');
+		sign.update(data);
+		sign.end();
+		const sig = sign.sign(Config.privatekeys[challengekeyid], 'hex');
+		return data + ';' + sig;
+	}
+	static isUseridAllowed(userid: string) {
+		const disallowed = [
+			...(Config.bannedTerms || []),
+			'nigger', 'nigga', 'faggot',
+			'lolicon', 'roricon', 'lazyafrican',
+			'tranny',
+		];
+		for (const term of disallowed) {
+			if (userid.includes(term)) return false;
+		}
+		return true;
+	}
+	static wordfilter(user: string) {
+		if (!Config.filteredterms) Config.filteredterms = [/(lol|ror)icon/];
+		for (const term of Config.filteredterms) {
+			user = user.replace(term, '*');
+		}
+		return user;
+	}
+	async changePassword(name: string, pass: string) {
+		const userid = toID(name);
+
+		const userData = await users.get('*', userid);
+		if (!userData) return false;
+
+		const entry = "Password changed from: " + userData.passwordhash;
```

Comment:
Does the usermodlog currently contain old password hashes? I thought it just said the user ID that changed the password? The usermodlog can be seen by US without access to MySQL...

Reply:
```php
   while ($row = $psdb->fetch_assoc($usermodlog)) {
        $entry = $row['entry'];
        $fromindex = strpos($entry, " from: ");
        if ($fromindex !== false) $entry = substr($entry, 0, $fromindex);

```
The hash is removed in usermodlog.php (posting for posterity). Not sure if the logging code needs to be retained, though.